From f42c7207625b977006d884bcd6491d0984d05e8c Mon Sep 17 00:00:00 2001
From: Hardik Shah <hardik.shah@ti.com>
Date: Mon, 16 Feb 2009 16:47:48 +0530
Subject: [PATCH 297/304] Source color keying support in DSS library

Tested only with LCD manager.
Only enabling part not the key part
---
 arch/arm/plat-omap/dss/dpi.c              |    9 +++++++++
 arch/arm/plat-omap/dss/dss.h              |    4 ++--
 arch/arm/plat-omap/dss/venc.c             |    9 +++++++++
 arch/arm/plat-omap/include/mach/display.h |    8 ++++++++
 4 files changed, 28 insertions(+), 2 deletions(-)

diff --git a/arch/arm/plat-omap/dss/dpi.c b/arch/arm/plat-omap/dss/dpi.c
index d749f0c..d8a197c 100644
--- a/arch/arm/plat-omap/dss/dpi.c
+++ b/arch/arm/plat-omap/dss/dpi.c
@@ -344,6 +344,14 @@ static enum omap_dss_update_mode dpi_display_get_update_mode(
 		OMAP_DSS_UPDATE_DISABLED;
 }
 
+static void dpi_display_set_color_keying(struct omap_display *display,
+	struct omap_color_key *key)
+{
+	omap_dispc_set_trans_key(OMAP_DSS_CHANNEL_LCD, key->type, key->color);
+	omap_dispc_enable_trans_key(OMAP_DSS_CHANNEL_LCD, key->enable);
+	dispc_go(OMAP_DSS_CHANNEL_LCD);
+}
+
 void dpi_init_display(struct omap_display *display)
 {
 	DSSDBG("DPI init_display\n");
@@ -359,6 +367,7 @@ void dpi_init_display(struct omap_display *display)
 	display->get_update_mode = dpi_display_get_update_mode;
 	display->set_bg_color = dpi_display_set_bg_color;
 	display->get_bg_color = dpi_display_get_bg_color;
+	display->set_color_keying = dpi_display_set_color_keying;
 }
 
 int dpi_init(void)
diff --git a/arch/arm/plat-omap/dss/dss.h b/arch/arm/plat-omap/dss/dss.h
index 9c2b281..4bc073b 100644
--- a/arch/arm/plat-omap/dss/dss.h
+++ b/arch/arm/plat-omap/dss/dss.h
@@ -229,10 +229,10 @@ void dispc_set_tft_data_lines(int data_lines);
 void dispc_set_lcd_display_type(enum omap_lcd_display_type type);
 void dispc_set_loadmode(enum omap_dss_load_mode mode);
 
-void dispc_set_trans_key(enum omap_channel ch,
+void omap_dispc_set_trans_key(enum omap_channel ch,
 		enum omap_dss_color_key_type type,
 		u32 trans_key);
-void dispc_enable_trans_key(enum omap_channel ch, int enable);
+void omap_dispc_enable_trans_key(enum omap_channel ch, int enable);
 
 void dispc_set_lcd_timings(struct omap_video_timings *timings);
 unsigned long dispc_fclk_rate(void);
diff --git a/arch/arm/plat-omap/dss/venc.c b/arch/arm/plat-omap/dss/venc.c
index 9da9bdc..aaf472e 100644
--- a/arch/arm/plat-omap/dss/venc.c
+++ b/arch/arm/plat-omap/dss/venc.c
@@ -514,6 +514,14 @@ static int venc_display_get_bg_color(struct omap_display *display)
 	return omap_dispc_get_default_color(OMAP_DSS_CHANNEL_DIGIT);
 }
 
+static void venc_display_set_color_keying(struct omap_display *display,
+	struct omap_color_key *key)
+{
+	omap_dispc_set_trans_key(OMAP_DSS_CHANNEL_DIGIT, key->type, key->color);
+	omap_dispc_enable_trans_key(OMAP_DSS_CHANNEL_DIGIT, key->enable);
+	dispc_go(OMAP_DSS_CHANNEL_DIGIT);
+}
+
 void venc_init_display(struct omap_display *display)
 {
 	display->panel = &venc_panel;
@@ -522,4 +530,5 @@ void venc_init_display(struct omap_display *display)
 	display->get_timings = venc_get_timings;
 	display->set_bg_color = venc_display_set_bg_color;
 	display->get_bg_color = venc_display_get_bg_color;
+	display->set_color_keying = venc_display_set_color_keying;
 }
diff --git a/arch/arm/plat-omap/include/mach/display.h b/arch/arm/plat-omap/include/mach/display.h
index 1ee41c0..3a319d1 100644
--- a/arch/arm/plat-omap/include/mach/display.h
+++ b/arch/arm/plat-omap/include/mach/display.h
@@ -284,6 +284,12 @@ struct omap_video_timings {
 
 };
 
+struct omap_color_key {
+	enum omap_dss_color_key_type type;
+	u32 color;
+	u32 enable;
+};
+
 struct omap_panel {
 	struct module *owner;
 
@@ -438,6 +444,8 @@ struct omap_display {
 			struct omap_video_timings *timings);
 	void (*get_timings)(struct omap_display *display,
 			struct omap_video_timings *timings);
+	void (*set_color_keying)(struct omap_display *display,
+			struct omap_color_key *key);
 	int (*update)(struct omap_display *display,
 			       int x, int y, int w, int h);
 	int (*sync)(struct omap_display *display);
-- 
1.5.6

